sdk/storage/storage-blob/review/storage-blob.api.ts

```diff
@@ -0,0 +1,1424 @@
+  // WARNING: The type "*     Models.ContainerListBlobHierarchySegmentResponse" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.BlobPrefix" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.BlobItem" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.ContainerListBlobHierarchySegmentResponse" needs to be exported by the package (e.g. added to index.ts)
+  listBlobsByHierarchy(delimiter: string, options?: ContainerListBlobsOptions): PagedAsyncIterableIterator<{
+          kind: "prefix";
+      } & Models.BlobPrefix | {
+          kind: "blob";
+      } & Models.BlobItem, Models.ContainerListBlobHierarchySegmentResponse>;
+  // WARNING: The type "Models.BlobItem" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.ContainerListBlobFlatSegmentResponse" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.BlobItem" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.ContainerListBlobFlatSegmentResponse" needs to be exported by the package (e.g. added to index.ts)
+  listBlobsFlat(options?: ContainerListBlobsOptions): PagedAsyncIterableIterator<Models.BlobItem, Models.ContainerListBlobFlatSegmentResponse>;
+  // WARNING: The type "Models.PublicAccessType" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.ContainerSetAccessPolicyResponse" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.PublicAccessType" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.ContainerSetAccessPolicyResponse" needs to be exported by the package (e.g. added to index.ts)
+  setAccessPolicy(access?: Models.PublicAccessType, containerAcl?: SignedIdentifier[], options?: ContainerSetAccessPolicyOptions): Promise<Models.ContainerSetAccessPolicyResponse>;
+  // WARNING: The type "Metadata" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.ContainerSetMetadataResponse" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Metadata" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.ContainerSetMetadataResponse" needs to be exported by the package (e.g. added to index.ts)
+  setMetadata(metadata?: Metadata, options?: ContainerSetMetadataOptions): Promise<Models.ContainerSetMetadataResponse>;
+  // WARNING: The type "Models.BlockBlobUploadResponse" needs to be exported by the package (e.g. added to index.ts)
+  // WARNING: The type "Models.BlockBlobUploadResponse" needs to be exported by the package (e.g. added to index.ts)
+  uploadBlockBlob(blobName: string, body: HttpRequestBody, contentLength: number, options?: BlockBlobUploadOptions): Promise<{
```

Comment:
HttpRequestBody is an odd type, it doesn't really feel like I can just pass a string for it (but you can).

Likewise, one of the things an HttpRequestBody can be is a function which returns a readable stream. It's odd that 1) I have to return a stream from a function rather than just give the stream directly, and 2) I need to know the size of the stream too. Would be nice if I could just pass the readable stream and the right things happen? I suppose it would have to load everything into memory in order to determine the size - maybe ok for small files, but a foot gun for large ones?

Reply:
Jeff points out that the callback lets you reset to the beginning of the stream. This does not seem obvious, and users might create the stream once and return it.

Reply:
Krzysztof suggestion: if you pass in a stream, AND you need to retry (or seek to beginning), we can throw.